### DISCUSSION TITLE
Analyze nullable return value statically?

### DISCUSSION THREAD

**Discussion by Cyl18** [dotnet/runtime#46496] [Q&A] [answered]
Consider this code:
```csharp
public static object? ToObject<T>(this T obj) => obj;
```

Is it possible to analyze whether the return value is nullable by the caller?
```csharp
string obj1 = "";
var obj1Result = obj1.ToObject(); // should be object

string obj2 = null;
var obj2Result = obj2.ToObject(); // should be object?
```

In Kotlin, we have code contract like this:
![image](https://user-images.githubusercontent.com/14993992/103436214-13516f80-4c54-11eb-993e-6eb5834e9ac8.png)
![image](https://user-images.githubusercontent.com/14993992/103436253-704d2580-4c54-11eb-9d46-c5f1b3fdec58.png)

(sorry for my bad grammar

**Comment by PathogenDavid** [comment, marked as answer]
I believe what you're looking for is the [nullable analysis attributes](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/attributes/nullable-analysis).

Specifically, `NotNullIfNotNull`: ([Full example on SharpLab here](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8kAmIA1AD4DEAdgK4A2tEwtcABHJY8wLABQAAgAYWfAIwA6ACKYIAc0oB7WJgDGUMQGF5BOAEEOtXFExQA3L158AzMJEA2YQCYWAFTixeAb14sfLANp8AOwgLABy8jChdLQAktjhkdEAFABE8sAAVikAlAC63r5WNvbpGXDKMAD8LvIA8pnlMAA8zgB8STAAFsYuLKXZLAC8rX2ZZjy+LAU+RaIlDRUubjAA+glR9EnZ0yxeE5OFIgLVpSJDLCkp4weFgaMZ4s51CzBbJiwA9B9h8iwA7hAEJRMJQZCwjJRlKwEG46DAWD0ANYKP6UFgwX7AVgKeE0eg7AC+Ox2szs90aS1ga2S232Pj2NxmRxOmTOgxYeNo10ZQXuj2eZQqb0+3wA6oDgaDwSCoSwYVA4QioCxkfJUejMdjooTeASgA))

```csharp
[return: NotNullIfNotNull("obj")]
public static object? ToObject<T>(this T obj) => obj;
```

--------

As an aside:

```csharp
var obj1Result = obj1.ToObject(); // should be object
```

With `var` this will always be `object?`, but the compiler will know it is not null. (Whether a particular variable is known to be null or able to be null are separate concepts. `var` is always nullable, but may or may not be known to be null.)

**Reply by Cyl18** [reply]
thanks!

**Reply by PathogenDavid** [reply]
No problem, happy new year!
